cmd/google-cloud-dns-healthcheck/run.go

```diff
@@ -0,0 +1,173 @@
+		err := fmt.Errorf("Wrong rrdatas value")
+		log.Errorf(err, fmt.Sprintf("No intersection between given rrdatas %v and record's rrdatas %v", r.Rrdatas, record.Rrdatas))
+		return err
+	}
+
+	healthyRrdatas := r.checkRrdatas()
+
+	switch len(healthyRrdatas) {
+	case len(record.Rrdatas):
+		log.Debugf("All rrdatas are healthy")
+		return nil
+	case 0:
+		log.Warningf("All rrdatas are unhealthy. We won't touch the record")
+		return nil
+	default:
+		log.Debugf("Updating record")
+		if err := r.updateDnsRecord(dnsService, record, healthyRrdatas); err != nil {
+			log.Errorf(err, "Error changing DNS RecordSet")
+			return err
+		}
+		return nil
+	}
+
+}
+
+func (r *RunCommand) getDnsService(ctx context.Context) (*dns.Service, error) {
+	c, err := google.DefaultClient(ctx, dns.CloudPlatformScope)
+	if err != nil {
+		return nil, err
+	}
+
+	dnsService, err := dns.New(c)
+	if err != nil {
+		return nil, err
+	}
+
+	return dnsService, nil
+}
+
+func (r *RunCommand) getDnsRecord(dnsService *dns.Service) (*dns.ResourceRecordSet, error) {
+	resp, err := dnsService.ResourceRecordSets.List(
+		r.Project, r.ManagedZone).Name(r.RecordName).Type(ResourceRecordSetTypeA).Do()
+
+	if err != nil {
+		return nil, fmt.Errorf("Record not found: %s", r.RecordName)
+	}
+
+	if len(resp.Rrsets) > 1 {
+		return nil, fmt.Errorf("Only expected 1 record set, got %d", len(resp.Rrsets))
+	}
+
+	return resp.Rrsets[0], nil
+}
+
+func (r *RunCommand) checkRrdatas() []string {
+	var client = &http.Client{
+		Timeout: time.Second * time.Duration(r.HttpTimeout),
+	}
+
+	healthyRrdatas := []string{}
+	for _, ip := range r.Rrdatas {
+		u := &url.URL{
+			Scheme: r.HttpScheme,
+			Host:   ip,
+			Path:   r.HealthcheckPath,
+		}
+
+		res, err := client.Get(u.String())
+		if err != nil {
+			log.Warningf("Error probing for %s", u.String())
+			continue
+		}
+
+		if res.StatusCode >= http.StatusOK && res.StatusCode < http.StatusBadRequest {
+			log.Infof("Success probing for %s", u.String())
+			healthyRrdatas = append(healthyRrdatas, ip)
+		} else {
+			log.Warningf("Error probing for %s: %d code", u.String(), res.StatusCode)
+		}
+
+	}
+
+	return healthyRrdatas
+}
+
+func (r *RunCommand) updateDnsRecord(dnsService *dns.Service, record *dns.ResourceRecordSet, rrdatas []string) error {
+	change := &dns.Change{
+		Deletions: []*dns.ResourceRecordSet{record},
+		Additions: []*dns.ResourceRecordSet{
+			{
+				Name:    record.Name,
+				Type:    record.Type,
+				Ttl:     record.Ttl,
+				Rrdatas: rrdatas,
+			},
+		},
+	}
+
+	log.Infof("DNS Record change request for %s - old: %v new: %v", record.Name, change.Deletions[0].Rrdatas, change.Additions[0].Rrdatas)
+
+	if !r.DryRun {
+		_, err := dnsService.Changes.Create(r.Project, r.ManagedZone, change).Do()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *RunCommand) intersection(a []string) []string {
+	intersection := []string{}
+	hash := make(map[string]bool)
+
+	for _, str := range r.Rrdatas {
```

Comment:
nit: `str` is quite generic while the same code uses `.Rrdatas`

Reply:
fixed in 03dff27